src/tags/object/RichText/domManager.ts

```diff
@@ -0,0 +1,624 @@
+    return dynamicBlock;
+  }
+  setDisplayedText(displayedText: string) {
+    this.displayedText = displayedText;
+  }
+  addStaticElement(currentNode: HTMLElement, path:Path) {
+    this.elements.push(new DDStaticElement(currentNode, this.endPos, path.toString()));
+  }
+  addExtraText(text:DDExtraText) {
+    let lastIdxOfTextBlock = this.elements.length - 1;
+
+    while (!(this.elements[lastIdxOfTextBlock] instanceof DDDynamicBlock) && lastIdxOfTextBlock > -1) {
+      --lastIdxOfTextBlock;
+    }
+    this.elements.splice(lastIdxOfTextBlock + 1, 0, text);
+  }
+  addTextElement(textNode: Text, path:Path) {
+    const { displayedText } = this;
+    const text:string = textNode.textContent as string;
+    const length = [...(text)].length;
+    const pos = displayedText.indexOf(text, this.displayedTextPos);
+
+    if (pos !== this.displayedTextPos) {
+      this.addExtraText(this.displayedText.substring(this.displayedTextPos, pos));
+      this.displayedTextPos = pos;
+    }
+    const dynamicBlock = this.createDynamicBlock(path.toString());
+
+    dynamicBlock.addTextNode(textNode, this.endPos, this.endPos + length, text, path.toString());
+    this.endPos += length;
+    this.displayedTextPos += text.length;
+  }
+  // That's mostly for processing html representation of simple text
+  // but historically we calculating this as one symbol even in html
+  // so we should keep it here anyway
+  addBR() {
+    this.endPos += 1;
+  }
+  findTextElement(pos:number, avoid: 'start'|'end' = 'start'):DDTextElement|undefined {
+    return this.findTextBlock(pos, avoid)?.findTextElement(pos, avoid);
+  }
+  findElementByPath(path: string) {
+    for (const el of this.elements) {
+      if (typeof el !== 'string' && el.path === path) {
+        return el;
+      }
+    }
+  }
+  findElementByNode(node: Node) {
+    for (const el of this.elements) {
+      if (el instanceof DDStaticElement) {
+        if (el.node === node) {
+          return el;
+        }
+      } else if (el instanceof DDDynamicBlock) {
+        const res = el.findElementByNode(node);
+
+        if (res) return res;
+      }
+    }
+  }
+  findTextBlock(pos:number, avoid: 'start'|'end' = 'start'):DDDynamicBlock|undefined {
+    const block = this.elements.find(el => {
+      return (el instanceof DDDynamicBlock) && el.start <= pos && el.end >= pos && el[avoid] !== pos;
+    });
+
+    return isDefined(block) ? block as DDDynamicBlock : block;
+  }
+  indexOfTextBlock(pos:number, avoid: 'start'|'end' = 'start'):number {
+    return this.elements.findIndex(el => {
+      return (el instanceof DDDynamicBlock) && el.start <= pos && el.end >= pos && el[avoid] !== pos;
+    });
+  }
+  getText(start: number, end: number) {
+    const startIdx = this.indexOfTextBlock(start, 'end');
+    const endIdx = this.indexOfTextBlock(end, 'start');
+
+    return this.elements.slice(startIdx, endIdx + 1).map(el => {
+      return typeof el !== 'string' ? el.getText(start, end) : el;
```

Comment:
We are sure that all other elements within the DOM representation, when calling their getText, that it returns a primitive value guaranteed?

Reply:
I think we can guarantee it as all the elements here are instances of `DDDynamicBlock` wich in the end just makes concatenation of all values returned by all descendants that are `DDTextElement` instances which just slicing its string represetantion collected before.
And I think I now understand exactly what I should describe about `domManager`.